pkg/controller/operators/internal/alongside/alongside.go

```diff
@@ -0,0 +1,80 @@
+// Package alongside provides a mechanism for recording the fact that
+// one object was installed alongside another object as part of the
+// installation of the same operator version.
+package alongside
+
+import (
+	"fmt"
+	"hash/fnv"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	prefix = "operatorframework.io/installed-alongside-"
```

Comment:
I recall we already have labels attached to CRD a part of Operator API. So does this work improve anything further as I feel this is a bit redundant?

Reply:
This associates a resource with a specific bundle installation, not a package, so that we can ensure that an owned CRD requirement is not satisfied until that step of its bundle installation has been applied.

Reply:
I think this is what you are already doing but I think we can simply add this prefix as key and value as a hash value of either CSV name or CSV name + version. When we are checking for requirement, we can compare that hash value to ensure the resources are what they supposed to be (either created/upgraded by the installplan). I would like to change the key a bit. Maybe `"operatorframework.io/install-id` if we end up using hash value.

Reply:
That's not sufficient because the same object can have multiple annotations of this form. This would be the case during an upgrade from one CSV to another providing the same CRD, or when two CSVs providing the same CRD are installed in two namespaces.

The value is "namespace/name" in order to be able to enumerate all of these CSVs. It would be possible to omit the value and instead list all CSVs on the cluster, then hash them and compare against the annotation key suffixes, but that would be sensitive to things like fixes for hash function bugs, and it's not possible in that case to regenerate these annotations once the InstallPlan has been garbage collected. It would also be possible to do no hashing at all and serialize a list of CSV names into the value of the annotation.

Reply:
I see what you meant. However, I'm not particularly a big fan of adding more and more annotations/labels into CSV and cascading all CSV names in that annotations. It only makes the in-cluster CSV becomes more bloated than it already is. I find massive CSV is just a painful to debug when it comes to it. Anyway, it is in-cluster representation so I don't think it is a big deal. I'm fine with it.
One more thing though if the descoping operator comes true, then we probably don't need worry about two instances of the same operator (owning the same CRD) anymore.

Reply:
The annotations aren't placed on CSVs, they're placed on things that were installed as part of the same bundle as a CSV.

I agree with that last point. I mention in the commit message that this is a stopgap measure to deal with the lack of a cluster-scoped API that represents the applied bundle itself. That's ideal, but it's not something that can be delivered as a bugfix for users that are seeing this problem today.